### ISSUE TITLE
Recursion error with version 0.4

### ISSUE THREAD

**Issue by angelo-romano** [smn/django-dirtyfields#22]
I am getting a maximum recursion error on a unit test after the switch to version 0.4. I assume it is being caused by the call to ._full_dict(), namely.

```python
...
  File "/home/angelo/.virtualenvs/online/src/django/django/db/models/query.py", line 228, in iterator
    obj = model_cls(**dict(zip(init_list, row_data)))
  File "/home/angelo/.virtualenvs/online/local/lib/python2.7/site-packages/dirtyfields/dirtyfields.py", line 14, in __init__
    reset_state(sender=self.__class__, instance=self)
  File "/home/angelo/.virtualenvs/online/local/lib/python2.7/site-packages/dirtyfields/dirtyfields.py", line 56, in reset_state
    instance._original_state = instance._full_dict()
  File "/home/angelo/.virtualenvs/online/local/lib/python2.7/site-packages/dirtyfields/dirtyfields.py", line 17, in _full_dict
File "/home/angelo/.virtualenvs/online/src/django/django/forms/models.py", line 143, in model_to_dict
    data[f.name] = f.value_from_object(instance)
  File "/home/angelo/.virtualenvs/online/src/django/django/db/models/fields/__init__.py", line 560, in value_from_object
    return getattr(obj, self.attname)
  File "/home/angelo/.virtualenvs/online/src/django/django/db/models/query_utils.py", line 114, in __get__
    instance._state.db).get(pk=instance.pk),
  File "/home/angelo/.virtualenvs/online/src/django/django/db/models/query.py", line 304, in get
    num = len(clone)
  File "/home/angelo/.virtualenvs/online/src/django/django/db/models/query.py", line 77, in __len__
    self._fetch_all()
  File "/home/angelo/.virtualenvs/online/src/django/django/db/models/query.py", line 857, in _fetch_all
    self._result_cache = list(self.iterator())
  File "/home/angelo/.virtualenvs/online/src/django/django/db/models/query.py", line 228, in iterator
    obj = model_cls(**dict(zip(init_list, row_data)))
  File "/home/angelo/.virtualenvs/online/local/lib/python2.7/site-packages/dirtyfields/dirtyfields.py", line 14, in __init__
    reset_state(sender=self.__class__, instance=self)
  File "/home/angelo/.virtualenvs/online/local/lib/python2.7/site-packages/dirtyfields/dirtyfields.py", line 56, in reset_state
    instance._original_state = instance._full_dict()
  File "/home/angelo/.virtualenvs/online/local/lib/python2.7/site-packages/dirtyfields/dirtyfields.py", line 17, in _full_dict
    return model_to_dict(self)
  File "/home/angelo/.virtualenvs/online/src/django/django/forms/models.py", line 143, in model_to_dict
    data[f.name] = f.value_from_object(instance)
  File "/home/angelo/.virtualenvs/online/src/django/django/db/models/fields/__init__.py", line 560, in value_from_object
    return getattr(obj, self.attname)
...
```

**Comment by romgar** [comment]
Can you post some code related to your test (and everything you think useful), so that I can try to reproduce it ?

**Comment by byashimov** [comment]
Same here.
`_full_dict` calls `model_to_dict(self)`
which calls obj's `__init__`
which calls `reset_state`
which calls `instance._full_dict()`

**Comment by romgar** [comment]
I'm at the end removing this `model_to_dict`, fix coming soon.

**Comment by romgar** [comment]
Can you give me some code to reproduce this problem, so that I can also create a test to be sure that my fix is correcting your problem @byashimov ?

**Comment by byashimov** [comment]
Here is your test case:

```python
def test_deferred_query(self):
    TestModel.objects.create()
    tm = TestModel.objects.only('boolean').first()
    self.assertEqual(tm.boolean, True)
```
```
RuntimeError: maximum recursion depth exceeded
----------------------------------------------------------------------
Ran 5 tests in 0.050s
FAILED (errors=1)
```

It doesn't work with deferred objects.

**Comment by romgar** [comment]
Thanks @byashimov !

**Comment by romgar** [comment]
Hi @angelo-romano , can you check with last released version, 0.4.1 ?

**Comment by angelo-romano** [comment]
it seems like not having fixed this, sorry.

/home/angelo/.virtualenvs/online/src/online-apps/myapp/models.py in __init__
        super(MyModel, self).__init__(*args, **kwargs) ...
▶ Local vars
/home/angelo/.virtualenvs/online/src/django/django/db/models/base.py in __init__
        super(Model, self).__init__() ...
▶ Local vars
/home/angelo/.virtualenvs/online/local/lib/python2.7/site-packages/dirtyfields/dirtyfields.py in __init__
        reset_state(sender=self.__class__, instance=self) ...
▶ Local vars
/home/angelo/.virtualenvs/online/local/lib/python2.7/site-packages/dirtyfields/dirtyfields.py in reset_state
    instance._original_state = instance._as_dict(check_relationship=True) ...
▶ Local vars
/home/angelo/.virtualenvs/online/local/lib/python2.7/site-packages/dirtyfields/dirtyfields.py in _as_dict
            all_field[field.name] = getattr(self, field.name) ...
▶ Local vars
/home/angelo/.virtualenvs/online/src/django/django/db/models/fields/related.py in __get__
                rel_obj = qs.get() ...
▶ Local vars
/home/angelo/.virtualenvs/online/src/django/django/db/models/query.py in get
        num = len(clone) ...
▶ Local vars
/home/angelo/.virtualenvs/online/src/django/django/db/models/query.py in __len__
        self._fetch_all() ...
▶ Local vars
/home/angelo/.virtualenvs/online/src/django/django/db/models/query.py in _fetch_all
            self._result_cache = list(self.iterator()) ...
▶ Local vars
/home/angelo/.virtualenvs/online/src/django/django/db/models/query.py in iterator
                    obj = model(*row_data) ...
▶ Local vars
/home/angelo/.virtualenvs/online/src/online-apps/myapp/models.py in __init__
        super(MyModel, self).__init__(*args, **kwargs)

**Comment by romgar** [comment]
@angelo-romano can you give some code related to this issue please ? Your test that is triggering this recursion error for example ?

**Comment by angelo-romano** [comment]
it's not triggered by a test, but by the app itself when trying to import the model. Will try to come around with a unit test for it.

**Comment by angelo-romano** [comment]
here we have it: https://github.com/smn/django-dirtyfields/pull/28

**Comment by romgar** [comment]
Thanks a lot @angelo-romano. Will work on it.

**Comment by romgar** [comment]
This issue is now fixed. You can test it on your project with `develop` branch, and will be released on `0.6` version.
